subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java

```diff
@@ -258,7 +273,24 @@ public void useJUnitJupiter(Provider<String> version) {
 
     @Override
     public void useSpock() {
-        useSpock(Frameworks.SPOCK.defaultVersion);
+        useSpockForDerivedGroovyVersion();
+    }
+
+    private void useSpockForDerivedGroovyVersion() {
+        final Set<String> groovyGroups = new HashSet<>(Arrays.asList("org.codehaus.groovy", "org.apache.groovy"));
+        final Set<String> groovyNames = new HashSet<>(Arrays.asList("groovy", "groovy-all"));
+        Provider<VersionNumber> groovyVersionProvider = getProviderFactory().provider(() -> {
+            Optional<VersionNumber> maybeGroovyJarVersionNumber = runtimeClasspath.getIncoming().getResolutionResult().getAllComponents().stream()
+                .map(ResolvedComponentResult::getModuleVersion)
+                .filter(mvi -> groovyGroups.contains(mvi.getGroup()))
+                .filter(mvi -> groovyNames.contains(mvi.getName()))
+                .map(ModuleVersionIdentifier::getVersion)
+                .map(VersionNumber::parse)
+                .findFirst();
+
+            return maybeGroovyJarVersionNumber.orElse(VersionNumber.parse(GroovySystem.getVersion()));
+        });
+        useSpock(groovyVersionProvider.map(groovyVersion -> String.format("%s-groovy-%s.%s", SPOCK_BASE_VERSION, groovyVersion.getMajor(), groovyVersion.getMinor())));
```

Comment:
Good question.

Spock 2.x only puts out binaries for Groovy 2.5, 3.0 and 4.0: https://repo1.maven.org/maven2/org/spockframework/spock-core/

So in this case, the user cannot say `useSpock()`; they'd have to specify `useSpock('1.3-groovy-2.4')`.

Reply:
Should we detect this case and warn the user instead of putting the wrong version of spock on the classpath? In which case, couldn't we just put the right version on the classpath instead?

Reply:
Maybe, but it's a question of how far back in time we want to go.

| Detected Groovy version | Resulting Spock version |
|---------------------------|-------------------------|
| 1.6 | [0.5-groovy-1.6](https://repo1.maven.org/maven2/org/spockframework/spock-core/0.5-groovy-1.6/) |
| 1.7 | [0.6-groovy-1.7](https://repo1.maven.org/maven2/org/spockframework/spock-core/0.6-groovy-1.7/) |
| 1.8 | [0.7-groovy-1.8](https://repo1.maven.org/maven2/org/spockframework/spock-core/0.7-groovy-1.8/) |
| 2.0 | [1.1-groovy-2.0](https://repo1.maven.org/maven2/org/spockframework/spock-core/1.1-groovy-2.0/) |
| 2.3 | [1.1-groovy-2.3](https://repo1.maven.org/maven2/org/spockframework/spock-core/1.1-groovy-2.3/) |
| 2.4 | [1.3-groovy-2.4](https://repo1.maven.org/maven2/org/spockframework/spock-core/1.3-groovy-2.4/) |
| 2.5 | [2.2-groovy-2.5](https://repo1.maven.org/maven2/org/spockframework/spock-core/2.2-groovy-2.5/) |
| 3.0 | [2.2-groovy-3.0](https://repo1.maven.org/maven2/org/spockframework/spock-core/2.2-groovy-3.0/) |
| 4.0 | [2.2-groovy-4.0](https://repo1.maven.org/maven2/org/spockframework/spock-core/2.2-groovy-4.0/) |

So I think we could implement a compatibility table like this, but it'll add complexity. We could just continue to provide a reasonable default but not try to be so backwards compatible - in those obscure cases like Groovy 1.6, users could just provide their own version. WDYT?

Reply:
If we had a bunch of legacy code to maintain to keep support for these versions, I'd say we shouldn't support them but given all we need is a lookup table, I'd lean towards the table. Especially since it's not like 4.0 is going to be the last version of Groovy. Eventually we're going to need this or similar logic. 

Really, all we're doing is replacing a `String.format` with a `Map.get` and controlling this table in the code. I don't think that's very unreasonable.